src/libs/ReportActionsUtils.ts

```diff
@@ -631,6 +653,121 @@ function isNotifiableReportAction(reportAction: OnyxEntry<ReportAction>): boolea
+
+    messageItems.push({
+        kind: 'text',
+        content: `${verb} `,
+    });
+
+    const originalMessage = reportAction?.originalMessage as CommonOriginalMessage;
+    const targetAccountIDs: number[] = originalMessage?.targetAccountIDs ?? [];
+    const mentions = targetAccountIDs.map((accountID) => {
+        const personalDetail = allPersonalDetails?.[accountID];
+
+        if (personalDetail) {
+            const displayNameOrLogin =
+                LocalePhoneNumber.formatPhoneNumber(personalDetail.login ?? '') || personalDetail?.displayName ? personalDetail?.displayName : Localize.translateLocal('common.hidden');
+            return {content: `@${displayNameOrLogin}`, accountID};
+        }
+        return {content: '', accountID: 0};
+    });
+
+    const lastMention = mentions.pop();
+
+    if (mentions.length === 0) {
+        messageItems.push({
+            kind: 'userMention',
+            ...lastMention,
+        });
+    } else if (mentions.length === 1) {
+        messageItems.push(
+            {
+                kind: 'userMention',
+                ...mentions[0],
+            },
+            {
+                kind: 'text',
+                content: ` ${Localize.translateLocal('common.and')} `,
+            },
+            {
+                kind: 'userMention',
+                ...lastMention,
+            },
+        );
+    } else {
+        mentions.forEach((mention) => {
+            messageItems.push({
+                kind: 'userMention',
+                ...mention,
+                content: `${mention.content}, `,
+                isComma: true,
+            });
+        });
+
+        messageItems.push(
+            {
+                kind: 'text',
+                content: `${Localize.translateLocal('common.and')} `,
+            },
+            {
+                kind: 'userMention',
+                ...lastMention,
+            },
+        );
+    }
+
+    const roomName = originalMessage?.roomName;
+    if (roomName) {
+        const preposition = isInviteRoom ? ` ${Localize.translateLocal('workspace.invite.to')} ` : ` ${Localize.translateLocal('workspace.invite.from')} `;
+
+        messageItems.push(
+            {
+                kind: 'text',
+                content: preposition,
+            },
+            {
+                kind: 'roomReference',
+                roomName,
+                roomID: originalMessage?.reportID,
+                content: roomName,
+            },
+        );
+    }
+
+    return messageItems;
+}
+
+function getActionItemFragmentChanelLog(reportAction: OnyxEntry<ReportAction>) {
+    const messageItems: MessageActionItemChanelLog[] = getChannelLogMemberAction(reportAction);
+    let html = '<muted-text>';
+
+    messageItems.forEach((messageItem) => {
+        switch (messageItem.kind) {
+            case 'userMention':
+                html += `<mention-user accountID=${messageItem.accountID}></mention-user>${messageItem.isComma ? ', ' : ''}`;
```

Comment:
Honestly, I was thinking about rendering these things directly via React components, but I see that [we do not have any clean pure-React components](https://github.com/Expensify/App/blob/4952c39a6c2dd69663550bfaf11a29ee3c51943c/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L52) for building messages "programmatically".

It's fine as it is for now.